src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/Utilities/FilePathNormalizer.cs

```diff
@@ -44,13 +59,57 @@ public static string Normalize(string? filePath)
         }
 
         // Otherwise, create a new string from our normalized char buffer.
-        unsafe
+        return CreateString(normalizedSpan);
+    }
+
+    public static string GetDirectory(string filePath)
+    {
+        var filePathSpan = filePath.AsSpanOrDefault();
+        if (filePathSpan.IsEmpty)
         {
-            fixed (char* buffer = normalizedSpan)
-            {
-                return new string(buffer, 0, normalizedSpan.Length);
-            }
+            throw new ArgumentNullException(nameof(filePath));
         }
+
+        using var _1 = ArrayPool<char>.Shared.GetPooledArray(filePathSpan.Length, out var array);
+        var arraySpan = array.AsSpan(0, filePathSpan.Length);
+
+        var (start, length) = NormalizeCore(filePathSpan, array.AsSpan(0, filePathSpan.Length));
+        var normalizedSpan = arraySpan.Slice(start, length);
+
+        var lastSlashIndex = normalizedSpan.LastIndexOf('/');
+
+        var directoryNameSpan = lastSlashIndex >= 0
+            ? normalizedSpan[..(lastSlashIndex + 1)] // Include trailiing slash
+            : normalizedSpan;
+
+        return CreateString(directoryNameSpan);
+    }
+
+    public static bool FilePathsEquivalent(string? filePath1, string? filePath2)
+    {
+        var filePathSpan1 = filePath1.AsSpanOrDefault();
+        var filePathSpan2 = filePath2.AsSpanOrDefault();
+
+        if (filePathSpan1.IsEmpty)
+        {
+            return filePathSpan2.IsEmpty;
+        }
+        else if (filePathSpan2.IsEmpty)
+        {
+            return false;
+        }
+
+        using var _1 = ArrayPool<char>.Shared.GetPooledArray(filePathSpan1.Length, out var array1);
+        var arraySpan1 = array1.AsSpan(0, filePathSpan1.Length);
+        var (start1, length1) = NormalizeCore(filePathSpan1, arraySpan1);
+        ReadOnlySpan<char> normalizedSpan1 = arraySpan1.Slice(start1, length1);
+
+        using var _2 = ArrayPool<char>.Shared.GetPooledArray(filePathSpan2.Length, out var array2);
```

Comment:
nit: would refactoring this block into a separate function reduce redundancy? This block and the one above seem to be doing the same thing

Reply:
I'll take a look. It's a little tricky because the "using" can't be factored out, since they control the lifetime of the array that the ReadOnlySpan is pointing at, but it should mostly be possible.